### PR TITLE
Automatically get Ganon Boss Key when meeting goal. (No longer tied to LACS.)

### PR DESCRIPTION
**Pull request by Cuphat**
Demonstration: https://www.youtube.com/watch?v=rPEXP52sYbw

This adds new options for Ganon Boss Key: Medallions, Stones, Dungeons, and Tokens. When reaching the goal amount of the type selected, you are automatically rewarded the Ganon BK. This allows LACS to be returned to a normal check but still having the flexibility of the bridge and beating Ganon not being tied together. Ganon BK can still be placed on LACS if wanted.

The LACS condition settings have now been hidden from the GUI. They still exist, however, and can be utilized on the command line or in a plando.

### CODE REVIEW

File: ItemPool.py [hunk 1]
```diff
@@ -1314,6 +1314,12 @@ def get_pool_core(world):
     elif world.settings.shuffle_ganon_bosskey == 'vanilla':
         placed_items['Ganons Tower Boss Key Chest'] = 'Boss Key (Ganons Castle)'
 
+    if world.settings.shuffle_ganon_bosskey in ['stones', 'medallions', 'dungeons', 'tokens']:
+         placed_items['Gift from Sages'] = 'Boss Key (Ganons Castle)'
```
Comment (Zannick) [inline ItemPool.py:1318]: nit: too many spaces on that indent

File: SettingsList.py [hunk 2]
```diff
@@ -3140,8 +3144,26 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
             appear on the Light Arrow Cutscene.
+            
+            'Stones': Ganon's Castle Boss Key will be awarded
+            when reaching the target number of Spiritual Stones.
+            
+            'Medallions': Ganon's Castle Boss Key will be awarded
+            when reaching the target number of Medallions.
+                        
+            'Dungeons': Ganon's Castle Boss Key will be awarded
+            when reaching the target number of Dungeon Rewards.
```
Comment (Zannick) [inline SettingsList.py:3155]: This includes the 1 in Link's Pocket, right? Should that be noted here? (Should the minimum be 2?)

File: ItemPool.py (hunk 1 above)
Comment (Cuphat) [reply ItemPool.py:1318]: Easy enough to fix.

File: SettingsList.py (hunk 2 above)
Comment (Cuphat) [reply SettingsList.py:3155]: It does include the 1 in Link's Pocket, same as LACS Condition and Bridge Condition. Minimum of 2 would make sense, but it'd also make sense in those as well.

File: ASM/c/ganon_boss_key.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#include <stdbool.h>
+#include "get_items.h"
+#include "z64.h"
+
+int num_to_bits[16] = { 0, 1, 1, 2, 1, 2, 2, 3,
+                        1, 2, 2, 3, 2, 3, 3, 4 };
+
+// https://www.geeksforgeeks.org/count-set-bits-in-an-integer/
+unsigned int countSetBitsRec(unsigned int num)
```
[the remark this answers is not in the document]
Comment (Cuphat) [reply ASM/c/ganon_boss_key.c:9]: `__builtin_popcount` seems to break armips.
```build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(119) warning: Content of areas 80400000 and 80400034 overlap
Aborting.```

File: ASM/c/ganon_boss_key.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                    give_boss_key = true;
+                break;
+            case 4: // Tokens
+                if (z64_file.gs_tokens >= GANON_BOSS_KEY_CONDITION_COUNT)
+                    give_boss_key = true;
+                break;
+        }
+
+        if (give_boss_key) {
+            push_delayed_item(0x03);
```
[the remark this answers is not in the document]
Comment (Cuphat) [reply ASM/c/ganon_boss_key.c:49]: That's doable, but it'd also mean that getting the boss key and doing a Don't Save/Don't Continue would mean that you no longer would get the boss key until you actually reset the game. Maybe we should do some further testing to see if the boss key being given multiple times is something that could happen?

**Review by Zannick — APPROVED** [review]

**Pull request merged**